### ISSUE TITLE
Logical expressions do not enforce logical requirement

### ISSUE THREAD

**Issue by ericbf** [microsoft/TypeScript#54252]
# Bug Report

### 🔎 Search Terms

logical condition inference, reduce logical condition, reduce logical expression, reduce condition, logical expression

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about logical expressions

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligObwA+8KArhBAFCiSwLLpYBGeBxZlNdevUIAzeAAoAhOKjwAvHKq0I8AGSr4HeYoEQAlGo3TZkncr0GA3vXi3NAOgghSGABbwA9B-gBRGDCQYe3gAYSRaYBQAciw3QjR4UAxwWNcoTnh4qlQAWl0AfnoAXyA)

### 💻 Code

```ts
declare const a: string | null
declare const b: string | null

if (!(a == null && b == null) && !(a != null)) {
    b.length // Error. Couldn't this detect that b is non-null?
}
```

### 🙁 Actual behavior

`b` is still nullable

### 🙂 Expected behavior

`b` should be non-null. The logical expression relatively trivially reduces to `b != null`:

1. `!(a == null && b == null) && !(a != null)`
2. `(a != null || b != null) && (a == null)`
3. `a != null && a == null || b != null && a == null`
4. `false || b != null && a == null`
5. `b != null && a == null`

----

I edited to remove the extended use case, as that’s not really the point here. This issue is really about a single, reducible logical expression in one `if` statement.

**Comment by fatcerberus** [comment]
Similar to and/or duplicate of https://github.com/microsoft/TypeScript/issues/30869 and others.  [The compiler is not a proof solver.](https://github.com/microsoft/TypeScript/issues/52822#issuecomment-1434989002)

Per https://github.com/microsoft/TypeScript/issues/30869#issuecomment-482316955:
> It's easy enough to counterfactually reason through the 2-variable case, but generalizing this is [really quite difficult](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem) and doesn't come up often enough for us to want to start trying to writing heuristic algorithms for NP-complete problems.

**Comment by ericbf** [comment]
I can see how closing #30869 would probably resolve this issue, but I would not categorize this as a duplicate, or even similar to, that issue.

While that issue is about complex flow type analyses, e.g., knowing that `!(!a && !b)` implies that `a || b` should be truthy when `a` and `b` are both falsible, this issue is about a single conditional expression not being evaluated correctly.

Consider this:

```ts
declare const a: string | null
declare const b: string | null

if (!(a == null && b == null) && !(a != null)) {
    b.length // Error. Couldn't this detect that b is non-null?
}

if ((a != null || b !...
```
[Playground Link](https://www.typescriptlang.org/play?ts=5.0.4#code/CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligObwA+8KArhBAFCiSwLLpYBGeBxZlNdevUIAzeAAoAhOKjwAvHKq0I8AGSr4HeYoEQAlGo3TZkncr0GA3vXi3NAOgghSGABbwA9B-gBRGDCQYe3gAYSRaYBQAciw3QjR4UAxwWNcoTnh4qlQAWl0AfnoAXyFRCRl4UyU6CkotKt0DdQlZBWr9Kxs7DkdnEjdPbz8AoNDwiEiY+DiEpJTptIyslFyC4tKxCoblQ2xtdtrNSrMa5taT-XhrOwcnF3cvX39A4LCI6NSsubBU9KPl1bKQolYSbERQCBoBDkOrHA5nfaNK5dWw9O79B7eABySHgIGeMHWoIk9Quu3O7U6NzRfQGjxxeIJxSAA)

You can see that compiler starts considering `b` to be non-nullable only when the literal `false` is inputted in the expression, despite the compiler understanding that `a == null && a != null` is always false.

```ts
declare const a: string | null

if (a == null) {
    // a is `null`
    console.log(a)
}

if (a != null) {
    // a is `string`
    console.log(a)
}

if (a != null && a == null) {
    // a is `never`
...
```
[Playground Link](https://www.typescriptlang.org/play?ts=5.0.4#code/CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligObwA+8KArhBAFD2EBm8AFFPALxdW0QBKeAG968cfAD0k7PEJp4AAxp1FYicnRIIIAHQQkJDgPoBfRi3acAhLxWCR68dNnylBYiTUTEqNNr0DIygTcyZWDnhbPjp4ADI42R4Yh1EfF043ZRAANxAYbw0-AP1DYzNGF1YsgCMkAKgURQAaeBrqLDQACyRaYDaERWYoCDQQRQB+ek1MeFZeGzt+eMTOZPsgA)

Maybe the real issue is that a logical expression that narrows the type to never still evaluates to `boolean` instead of `false`?

**Comment by ericbf** [comment]
So this is correct, desired behavior?

```ts
declare const a: string | null

// Compiler understands this is equivalent to false
if (a != null && a == null) {
    // a is `never`
    console.log(a)
}

// type of `f` is literal `false`, so it does infer literal types for constants
const f = false

// type of `b` is `boolean`, despite compiler understanding that it is equivalent to false (illustrated above)
const b = a != null && a == null
```
[Playground link](https://www.typescriptlang.org/play?ts=5.0.4#code/CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligObwA+8KArhBAFD0D0T8AwkgLYAOhEIMeNRSgYmKCLTwMAC0JT58EAEdqhAG5R+KLBiTwAZlrQh6hA-AAUUeAEIAvFVoR4AMlfZ49xzToBKeABvenhQ+BZPRQADFBB1ASiQsOR0JH4AOggkEms-egBfRgiMAE9uBCQLKIMo+EUIQgwBLXhq4xAogBp8fUb4YCQQBRQDAXgGppgW0vKpAyRBFPEdNHolrAtHIwgTItYZiqqAI1roo6Q0kAku-qHeJsQuXn5BYVFl4GIyWSgsPsUVGpNNpdPptiYrHwINQCL8QMBsOd4nl1vAjl5PA4nHQ3B4bN5sRAgA)

Part of the issue here is that these two should be equivalent as far as the compiler is concerned:

```ts
declare const a: string | null
declare const b: string | null

if (a != null && a == null || b != null && a == null) {
    b.length // Error, but should be equivalent to the below
}

if (false || b != null && a == null) {
    b.length // No error
}
```
[Playground link](https://www.typescriptlang.org/play?ts=5.0.4#code/CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligObwA+8KArhBAFCiSwLLpYBGeBxZlNdevUIAzeAAoo8AIQBeKrQjwAZMuzxZ8gUvKUOMrYpVqpmhXQCU8AN7149+BwB0EEKQwALeAHpv8AKIwMEgwADSO1FhoHki0wI4IIACO1IQAblCuKFgYSPCeCBwgEEgA7vQAvkKiEiKZaAi6jgbmSqrqZtpWtg6OLm4knj5+AHJ5IEEhlUA)

**Comment by fatcerberus** [comment]
It doesn't actually understand that `a != null && a == null` is false - you've just narrowed away all possible types `a` could be so that `a` becomes typed as `never`.  Which happens in two separate steps - hover over the `a` after the `&&` to see what I mean.  The distinction is relevant because, if you actually replace the whole condition with `if (false) { ... }`, you get an unreachable code error on `console.log(a)`.

TS doesn't even attempt to reason about complex boolean expressions in the way you're imagining; instead, type narrowing applies to each operation independently.

**Comment by ericbf** [comment]
Gotcha. Is what it is then! Thanks!
